design/one-pager-generic-usage-type.md

```diff
@@ -0,0 +1,397 @@
+Crossplane, and is not deleted along with the EKS cluster. This means that the
+Helm Release (and therefore its Service) must be deleted before the EKS cluster
+in order to ensure the EKS cluster’s controllers trigger deletion of the ELB.
+If it is not, the ELB will be orphaned, costing money and blocking the deletion
+of the VPC it was created in.
+
+These problems are exacerbated by the fact that Kubernetes uses "background
+cascading deletion" by default. Put otherwise, the successful deletion of a
+parent resource is not dependent on the successful deletion of its children.
+Instead, the parent resource will be deleted successfully and the children
+(i.e. resources that declare the parent as an owner reference) will later be
+deleted by the Kubernetes garbage collector. This makes it difficult to diagnose
+or even notice resources that become orphaned when their parent XR is deleted.
+
+## Goals
+
+- Support "cross-provider" dependencies, for example an MR from provider-helm
+depending on a ProviderConfig from provider-aws.
+- Don’t require Crossplane users to manage fine-grained RBAC access.
+- Don’t grant Crossplane providers broad RBAC access (e.g. to all other providers).
+
+## Proposal
+
+We propose introducing a new `Usage` type that can be used to declare usage
+relationships between Crossplane resources. This type will be defined by the
+Core Crossplane repository, and be available for use as part of a composition.
+
+The relations defined by this type will be enforced by an admission webhook that
+will be running as part of the Crossplane core. Similar to the upstreams [liens
+proposal], the webhook will reject deletions of resources that are in use by
+other resources with a ["409 Conflict" error code].
+
+### API
+
+We will introduce a new `Usage` type as follows:
+
+```yaml
+apiVersion: apiextensions.crossplane.io/v1alpha1
+kind: Usage
+metadata:
+  name: release-uses-cluster
+spec:
+  of:
+  - apiVersion: eks.upbound.io/v1beta1
+    kind: Cluster
+    name: my-cluster
+  by:
+    apiVersion: helm.crossplane.io/v1beta1
+    kind: Release
+    name: my-prometheus-chart
+```
+
+The `spec.by` field will define the resource that will be using the resources
+defined in `spec.of`. Both will support only cluster-scoped resources, namely
+`Composites` and `Managed Resources`.
+
+In this example, we define a usage relationship between a Helm `Release` and an
+EKS `Cluster`. The deletion of the `Cluster` resource will be rejected as long
+as this `Release` resource exists.
+
+In addition to direct referencing with names, we will also support selectors to
+match labels or controller reference. This will allow us to define usage
+relationships between resources created by the same composition instance.
+
+```yaml
+apiVersion: apiextensions.crossplane.io/v1alpha1
+kind: Usage
+metadata:
+  name: release-uses-cluster
+spec:
+  of:
+  - apiVersion: eks.upbound.io/v1beta1
+    kind: Cluster
+    selector:
+      matchControllerRef: true
```

Comment:
This is asymmetrical with our current reference/selector implementations, right? It looks like a selector _inside_ a reference, rather than a selector _at the same level_ as a reference.

Would we ever have a use for different kinds of selectors - e.g. matchLabels? I think yes - e.g. if there are two `Cluster` in a Composition and we only want to select one (without knowing its name in advance) matching controller ref alone is insufficient.

Reply:
It might be useful to include an example of a Usage (or two) inside a Composition.

Reply:
> This is asymmetrical with our current reference/selector implementations, right? It looks like a selector inside a reference, rather than a selector at the same level as a reference.

Yes, the main difference is that for MRs, name/id refers to the external resource we don't have. For example:

```
apiVersion: ec2.aws.upbound.io/v1beta1
kind: Subnet
...
spec:
  forProvider:
    vpcId: <external-resource-id>
    vpcIdRef:
      name: <name-of-the-k8s-resource>
    vpcIdSelector:
      matchControllerRef: true
```

Here, the external name does not apply; hence we there is no reason to put the name under ref.

```
apiVersion: helm.crossplane.io/v1beta1
kind: Release
name: <name-of-the-k8s-resource
selector:
  matchControllerRef: true
  matchLabels:
    foo: bar
```

> Would we ever have a use for different kinds of selectors - e.g. matchLabels?

Yes. This would be part of the API while not included in the example. Updated.

Reply:
I'm a little wary of this API shape, which is kind of a-selector-inside-a-reference. Is it possible to stick to the existing pattern of selectors living next to references? My main concern is it seems to go against Kubernetes API conventions. To be fair our existing reference and selector pattern is based on old conventions (new ones are [here](https://github.com/kubernetes/community/blob/9885966e5015c3675b63116c268d7da0708b137d/contributors/devel/sig-architecture/api-conventions.md#object-references-examples)), but I'd prefer to be consistent with what we do elsewhere if possible.

I think we had a similar issue with EnvironmentConfigs and landed on this API (which to be fair I don't love either): https://github.com/crossplane/crossplane/pull/3008/files#diff-1894e5714a8f815a975c89cc55edb79c1c216dd5c1408c82ab84522d3acc3e83R94

Reply:
Ok, if I attempt to apply that pattern here, we would have one of the following:

```yaml
apiVersion: apiextensions.crossplane.io/v1alpha1
kind: Usage
metadata:
  name: release-uses-clusters
spec:
  of:
    - type: Reference
      reference:
        apiVersion: eks.upbound.io/v1beta1
        kind: Cluster
        name: my-cluster
    - type: Selector
      selector:
        apiVersion: eks.upbound.io/v1beta1
        kind: Cluster
        matchControllerRef: true
        matchLabels:
          foo: bar
  by:
    type: Reference
    reference:
      apiVersion: helm.crossplane.io/v1beta1
      kind: Release
      name: my-prometheus-chart
```

or

```yaml
apiVersion: apiextensions.crossplane.io/v1alpha1
kind: Usage
metadata:
  name: release-uses-clusters
spec:
  of:
    - apiVersion: eks.upbound.io/v1beta1
      kind: Cluster
      type: Reference # can't we just drop this type?
      reference:
        name: my-cluster
    - apiVersion: eks.upbound.io/v1beta1
      kind: Cluster
      type: Selector # can't we just drop this type?
      selector:
        matchControllerRef: true
        matchLabels:
          foo: bar
  by:
    apiVersion: helm.crossplane.io/v1beta1
    kind: Release
    type: Reference
    reference:
      name: my-prometheus-chart
```

TBH, I am not happy with either but closer to the latter.

This is slightly different from then EnvironmentConfig case, where the kind is known not arbitrary. Is this the first time where we are defining a multi-kind reference/selector type relation? And I believe we will need the same for generic resource references.

Reply:
> Is this the first time where we are defining a multi-kind reference/selector type relation?

I think so, yes.

```yaml
  of:
    - type: Reference
      reference:
        apiVersion: eks.upbound.io/v1beta1
        kind: Cluster
        name: my-cluster
    - type: Selector
      selector:
        apiVersion: eks.upbound.io/v1beta1
        kind: Cluster
        matchControllerRef: true
        matchLabels:
          foo: bar
```

This is my preference of the three. All of the APIs feel a little awkward to me though, so I leave it up to you to pick which you think is best. 😉

Reply:
Currently, I am closer to the second and `type` dropped but love to hear more thoughts:

```
  of:
    - apiVersion: eks.upbound.io/v1beta1
      kind: Cluster
      reference:
        name: my-cluster
    - apiVersion: eks.upbound.io/v1beta1
      kind: Cluster
      selector:
        matchControllerRef: true
        matchLabels:
          foo: bar
```

When the selector is configured, the resolution will be done and the controller change it from:

```
    - apiVersion: eks.upbound.io/v1beta1
      kind: Cluster
      selector:
        matchControllerRef: true
        matchLabels:
          foo: bar
```

to 

```
    - apiVersion: eks.upbound.io/v1beta1
      kind: Cluster
      reference:
        name: that-cluster
      selector:
        matchControllerRef: true
        matchLabels:
          foo: bar
```

By following the rules documented in the design:

> - Resolution will be made once when the `Usage` is created.
> - If multiple resources match the selector, a random one will be
>   selected.
> 
> The [existing policies] for resolving Managed Resource references does not
> necessarily apply to the `Usage`. For example, supporting an `optional`
> reference while defining a usage relationship does not make sense. Also, the
> [reasoning behind] supporting different policies for MRs does not apply to
> the `Usage`. However, depending on the feedback, we may consider supporting
> resolve policies to configure whether the resolution should be made once or
> continuously in a future iteration.

Reply:
We discussed this offline with @negz and aligned on the following API.

```yaml
apiVersion: apiextensions.crossplane.io/v1alpha1
kind: Usage
metadata:
  name: release-uses-clusters
spec:
  of:
    apiVersion: eks.upbound.io/v1beta1
    kind: Cluster
    resourceSelector:
      matchControllerRef: true
      matchLabels:
         foo: bar
  by:
    apiVersion: helm.crossplane.io/v1beta1
    kind: Release
    resourceRef:
      name: my-prometheus-chart
```

Given there is no solid reason to have `spec.of` as an array and multiple `Usage` objects could provide the same functionality, we wanted to drop it being an array due to potential confusion when a selector is provided, i.e. when a selector is provided, would all matching resources be added to the array?

The controller would resolve the references based on selectors, if any, and persist them into `resourceRef.name`, similar to how MR selectors are resolved today. So, the above would turn to the following after the first reconcile:

```yaml
apiVersion: apiextensions.crossplane.io/v1alpha1
kind: Usage
metadata:
  name: release-uses-clusters
spec:
  of:
    apiVersion: eks.upbound.io/v1beta1
    kind: Cluster
    resourceRef: # added by the controller based on the selectors
      name: my-cluster # added by the controller based on the selectors
    resourceSelector:
      matchControllerRef: true
      matchLabels:
         foo: bar
  by:
    apiVersion: helm.crossplane.io/v1beta1
    kind: Release
    resourceRef:
      name: my-prometheus-chart
```

Reply:
Updated the doc accordingly.